django/db/models/constraints.py

```diff
@@ -265,3 +289,61 @@ def deconstruct(self):
         if self.opclasses:
             kwargs["opclasses"] = self.opclasses
         return path, self.expressions, kwargs
+
+    def validate(self, model, instance, exclude=None, using=DEFAULT_DB_ALIAS):
+        queryset = model._default_manager.using(using)
+        if self.fields:
+            lookup_kwargs = {}
+            for field_name in self.fields:
+                if exclude and field_name in exclude:
+                    return
+                field = model._meta.get_field(field_name)
+                lookup_value = getattr(instance, field.attname)
+                if lookup_value is None or (
+                    lookup_value == ""
+                    and connections[using].features.interprets_empty_strings_as_nulls
+                ):
+                    # A composite constraint containing NULL value cannot cause
+                    # a violation since NULL != NULL in SQL.
+                    return
+                lookup_kwargs[field.name] = lookup_value
+            queryset = queryset.filter(**lookup_kwargs)
+        else:
+            # Ignore constraints with excluded fields.
+            if exclude:
+                for expression in self.expressions:
+                    for expr in expression.flatten():
+                        if isinstance(expr, F) and expr.name in exclude:
+                            return
+            replacement_map = instance._get_field_value_map(
+                meta=model._meta, exclude=exclude
+            )
+            expressions = [
+                Exact(expr, expr.replace_references(replacement_map))
+                for expr in self.expressions
+            ]
+            queryset = queryset.filter(*expressions)
+        model_class_pk = instance._get_pk_val(model._meta)
+        if not instance._state.adding and model_class_pk is not None:
+            queryset = queryset.exclude(pk=model_class_pk)
+        if not self.condition:
+            if queryset.exists():
+                if self.expressions:
+                    raise ValidationError(self.get_violation_error_message())
+                # When fields are defined, use the unique_error_message() for
+                # backward compatibility.
+                for model, constraints in instance.get_constraints():
+                    for constraint in constraints:
+                        if constraint is self:
+                            raise ValidationError(
+                                instance.unique_error_message(model, self.fields)
+                            )
+        else:
+            against = instance._get_field_value_map(meta=model._meta, exclude=exclude)
+            try:
+                if (self.condition & Exists(queryset.filter(self.condition))).check(
+                    against, using=using
+                ):
+                    raise ValidationError(self.get_violation_error_message())
+            except FieldError:
+                pass
```

Comment:
I've noticed that a field used in a condition might be excluded/missing in the `against` map, because it is excluded from the form (an internal status field).
The resulting `FieldError` is swallowed then.
Is this expected?  Why should anything get excluded from `against` here (and likely other places from this patch)?

Reply:
This is [the documented behaviour for forms](https://docs.djangoproject.com/en/4.1/topics/forms/modelforms/#interaction-with-model-validation)

> If you have excluded any model fields, validation will not be run on those fields. See the [form validation](https://docs.djangoproject.com/en/4.1/ref/forms/validation/) documentation for more on how field cleaning and validation work.

I guess the [form validation](https://docs.djangoproject.com/en/4.1/ref/forms/validation/) section could go into more details on the subject but this is how `unique_together` validation has been working forever

https://github.com/django/django/blob/08688bd7dd8dfa218ecff03d6c94a753b1be8e59/django/db/models/base.py#L1235-L1242

Reply:
The point is that I am not using `exclude` myself, and it is good that `status` is excluded from checks itself.
But the constraint is for `name`, and `status` is only used in the condition there:
```
    class Meta(…):
        constraints = [
            UniqueConstraint(fields=("name",), condition=Q(status="draft"), name="uniq_name_draft"),
        ]
 ```

Reply:
> It just happens to pretty straightforward here as you can directly call `Constraint.validate` without the `exclude` on the constraint you are interested in validating.

That's a suitable workaround, but I feel like it should not be necessary.
FWIW before Django 4.1 where this feature was added I added manual validation already, since there constraints with conditions where just skipped.